### PR TITLE
feat(optionality): add optionality components to units

### PR DESCRIPTION
**Pull request by benprotheroe**
## Description

Music year: {owa_music_year}

- List of changes

handle units with optionality
uses the optionality item compoents

### Implementation details

Plan:

- The query will return null univariants alongside their optionality counterparts
- In the query after parsing the schema:
    - Group null unitvariants with optionality variants
    - Where optionality accompanies a null variant:
        - filter out null unitvariants
        - use optionality listing item to present multiple options together
        - NB. it is possible to have a single optionality uv too
        - NB. we need to watch out for unitvariants that have programme factors but aren’t null (the MV might already do this)

### Acceptance criteria

- [x]  optionality unitvariants are correctly rendered using the oak optionality listing component
- [x]  all other unitvariants are uneffected

## Issue(s)

Fixes #

## How to test

1. Go to https://deploy-preview-2443--oak-web-application.netlify.thenational.academy/pupils/beta/programmes/english-secondary-year-10-edexcel/units
2. Click on optionality units
3. You should see all optionality units with correct links

## Screenshots

How it used to look (delete if n/a):
{screenshots}

How it should now look:
{screenshots}

## Checklist

- [ ] Added or updated tests where appropriate
- [ ] Manually tested across browsers / devices
- [ ] Considered impact on accessibility
- [ ] Design sign-off
- [ ] Approved by product owner
- [ ] Does this PR update a package with a breaking change

### CODE REVIEW

**Comment by messmike** [comment]
Do we need a seperate ticket for the update for optional units on the lesson listing page, or can we get it in as part of this one? E.g:
Screenshot 2024-05-22 at 11 37 21 am

**Comment by messmike** [comment]
@benprotheroe Rich found this example of what 5 optional units looks like: [https://deploy-preview-2443--oak-web-application.netlify.thenational.academy/pupils/beta/programmes/history-primary-year-1/units](https://deploy-preview-2443--oak-web-application.netlify.thenational.academy/pupils/beta/programmes/history-primary-year-1/units)

I think there is a bit of an alignment flex issue for the optional unit items when they wrap e.g:
Screenshot 2024-05-22 at 12 12 58 pm

Can they span equal column widths instead?

**Comment by RBrownsett** [comment]
@benprotheroe where is the count for the new lessons coming from? In the screen shot below it says 390 new lessons but when I add up all the lessons from the displayed units I only get 165? I do wonder if we should even show the lesson count at all - but thats a bigger question. 

Link to page tested https://deploy-preview-2443--oak-web-application.netlify.thenational.academy/pupils/beta/programmes/english-secondary-year-10-edexcel/units


image

**Comment by messmike** [comment]
> Do we need a seperate ticket for the update for optional units on the lesson listing page, or can we get it in as part of this one? E.g: Screenshot 2024-05-22 at 11 37 21 am
> 
> its a different repo so lets go different ticket for now

https://www.notion.so/oaknationalacademy/OWA-Handle-optionality-on-lessonlisting-page-22c30628e9504b2cb3728b72e56c6404?pvs=4

**Review by messmike — APPROVED** [review]

**Review by RBrownsett — APPROVED** [review]
Looks good! Lesson count now correct, mobile view displays correctly. Full testing details on the ticket.

File: src/components/PupilViews/PupilLessonListing/PupilLessonListing.view.tsx [hunk 1]
```diff
@@ -48,12 +48,17 @@ export const PupilViewsLessonListing = (props: PupilLessonListingViewProps) => {
     breadcrumb.push(tierDescription);
   }
 
+  console.log(programmeFields);
```
Comment (BTitterington) [inline src/components/PupilViews/PupilLessonListing/PupilLessonListing.view.tsx:51]: Accidental log.

**Review by BTitterington — APPROVED** [review]
Sorry for the delay. Apart from the accidental console.log it looks good.

**Pull request merged**